deeppavlov/core/models/tf_model.py

```diff
@@ -13,119 +13,97 @@
-from overrides import overrides
 
 from deeppavlov.core.models.nn_model import NNModel
-from deeppavlov.core.models.serializable import Serializable
-from deeppavlov.core.common.attributes import check_attr_true
 from deeppavlov.core.common.log import get_logger
-from deeppavlov.core.common.errors import ConfigError
 from .tf_backend import TfModelMeta
 
-"""
-Here is an abstract class for neural network models based on Tensorflow.
-If you use something different, ex. Pytorch, then write similar to this class, inherit it from
-Trainable and Inferable interfaces and make a pull-request to deeppavlov.
-"""
-
 log = get_logger(__name__)
 
 
-class TFModel(Serializable, metaclass=TfModelMeta):
-    def __init__(self, **kwargs):
-        self._saver = tf.train.Saver
-        super().__init__(**kwargs)
-
-    @abstractmethod
-    def _add_placeholders(self):
-        """
-        Add all needed placeholders for a computational graph.
-        """
-        pass
-
-    @abstractmethod
-    def run_sess(self, *args, **kwargs):
-        """
-        1. Call _build_graph()
-        2. Define all comuptations.
-        3. Run tf.sess.
-        3. Reset state if needed.
-        :return:
-        """
-        pass
-
-    @abstractmethod
-    def _train_step(self, features, *args):
-        """
-        Define a single training step. Feed dict to tf session.
-        :param features: input features
-        :param args: any other inputs, including target vector, you need to pass for training
-        :return: metric to return, usually loss
-        """
-        pass
-
-    @abstractmethod
-    def _forward(self, features, *args):
-        """
-        Pass an instance to get a prediction.
-        :param features: input features
-        :param args: any other inputs you need to pass for training
-        :return: prediction
-        """
-        pass
-
-    @check_attr_true('train_now')
-    def train(self, features, *args, **kwargs):
-        """
-        Just a wrapper for a private method.
-        """
-        return self._train_step(features, *args, **kwargs)
+class TFModel(NNModel, metaclass=TfModelMeta):
+    def __init__(self, *args, **kwargs):
+        if not hasattr(self, 'sess'):
+            raise RuntimeError('Your tensorflow model {} must'
+                               ' have sess attribute!'.format(self.__class__.__name__))
+        super().__init__(*args, **kwargs)
 
-    def infer(self, instance, *args, **kwargs):
-        """
-        Just a wrapper for a private method.
-        """
-        return self._forward(instance, *args, **kwargs)
+    def load(self):
+        """Load model parameters from self.load_path"""
+        path = str(self.load_path.resolve())
+        # Check presence of the model files
+        if tf.train.checkpoint_exists(path):
+            log.info('[loading model from {}]'.format(path))
+            # Exclude optimizer variables from saved variables
+            var_list = [var for var in tf.trainable_variables()
```

Comment:
Maybe we should make variables' exclusion optional (via method parameter?)
Or
```python
def load(self, exclude_scopes=['Optimizer']):
    <...>
    var_list = self.get_trainable_variables(exclude_scopes)
    <..>
def _get_trainable_variables(scope_names=[]):
       all_vars = tf.trainable_variables()
       if scope_names:
            vars_to_train = set()
            for sn in scope_names:
                scope_vars = tf.get_collection(tf.GraphKeys.TRAINABLE_VARIABLES, scope=sn)
                vars_to_train.update(scope_vars)
            return list(vars_to_train)
        else:
            return all_vars
```

Reply:
```
def load(self, exclude_scopes=['Optimizer']):
    <...>
    var_list = self.get_trainable_variables(exclude_scopes)
    <..>
def _get_trainable_variables(exclude_scopes=[]):
       all_vars = tf.trainable_variables()
       vars_to_train = [var for var in all_vars if all(sc not in var.name for sc in exclude_scopes)]
       return vars_to_train
```

Reply:
in progress

Reply:
Done